File: ai_houkai/cli/commands/ingest.py
"""Ingest command — chunk files (or stdin) into memories."""

from __future__ import annotations

import sys
from pathlib import Path
from typing import List, Optional

import typer

from ai_houkai.cli import output as out
from ai_houkai.memory_system.importance import score_importance
from ai_houkai.memory_system.ingest import chunk_text
from ai_houkai.memory_system.store import RememberItem


def ingest(
    ctx: typer.Context,
    files: Optional[List[Path]] = typer.Argument(
        None,
        help="Text/markdown files to ingest. Omit (or pass '-') to read stdin.",
    ),
    type: str = typer.Option("episodic", "-t", "--type", help="episodic|semantic|procedural|feedback"),
    tags: List[str] = typer.Option([], "-g", "--tag", help="Tag (repeatable)"),
    source: Optional[str] = typer.Option(
        None, "-s", "--source",
        help="Source label; default ingest:<filename> (or ingest:stdin)",
    ),
    importance: Optional[float] = typer.Option(None, "-i", "--importance", min=0.0, max=1.0),
    auto_importance: bool = typer.Option(
        False, "--auto-importance", help="Score each chunk heuristically"
    ),
    max_chars: int = typer.Option(500, "--max-chars", help="Max chunk size"),
    min_chars: int = typer.Option(30, "--min-chars", help="Drop chunks shorter than this"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Show chunks without writing"),
    yes: bool = typer.Option(False, "--yes", "-y"),
) -> None:
    """Split files into chunks and store each chunk as a memory.

    Splits on blank lines, keeps markdown headings attached to their
    paragraph, re-packs long paragraphs on sentence boundaries.
    """
    store = ctx.obj["store"]
    cfg = ctx.obj["config"]

    # (label, text) per input
    inputs: list[tuple[str, str]] = []
    from_stdin = not files or [str(f) for f in files] == ["-"]
    if from_stdin:
        body = sys.stdin.read()
        if not body.strip():
            typer.echo("Error: no input provided", err=True)
            raise typer.Exit(1)
        inputs.append(("stdin", body))
    else:
        for f in files:
            if not f.exists():
                typer.echo(f"Error: {f} not found", err=True)
                raise typer.Exit(1)
            inputs.append((f.name, f.read_text(encoding="utf-8", errors="replace")))

    auto = auto_importance or (
        importance is None and cfg.default_importance == "auto"
    )

    plan: list[tuple[str, str, float]] = []   # (label, chunk, importance)
    for label, body in inputs:
        for chunk in chunk_text(body, max_chars=max_chars, min_chars=min_chars):
            if importance is not None:
                imp = importance
            elif auto:
                imp = score_importance(chunk, type, list(tags))
            else:
                imp = (
                    cfg.default_importance
                    if isinstance(cfg.default_importance, float)
                    else 0.5
                )
            plan.append((label, chunk, imp))

    if not plan:
        typer.echo("Nothing to ingest (all chunks below --min-chars?).")
        return

    for i, (label, chunk, imp) in enumerate(plan, 1):
        first_line = chunk.splitlines()[0][:70]
        typer.echo(f"  [{i:3}] {imp:.2f}  {label}: {first_line}")
    typer.echo(f"\n{len(plan)} chunk(s) from {len(inputs)} input(s).")

    if dry_run:
        typer.echo("Dry-run — nothing written.")
        return

    if not out.confirm(f"Store {len(plan)} memories?", yes=yes, use_tty=from_stdin):
        typer.echo("Aborted.")
        return

    # friendly_errors: a bad --type/--tag surfaces as a one-line error, not a
    # traceback. Journal as "ingest" — "import" is the .ahkai importer's actor.
    # One batched write — collapses N per-chunk encodes into ceil(N/batch)
    # (see MemoryStore.remember_many). on_conflict="ignore": ingesting raw
    # document chunks shouldn't trigger conflict management (near-duplicate
    # chunks — shared headings/boilerplate — are normal source material).
    with store.as_actor("ingest"), out.friendly_errors():
        store.remember_many(
            [
                RememberItem(
                    text=chunk,
                    type=type,
                    tags=tuple(tags),
                    importance=imp,
                    source=source or f"ingest:{label}",
                )
                for label, chunk, imp in plan
            ],
            on_conflict="ignore",
        )
    typer.echo(f"Stored {len(plan)} memories.")

File: go/internal/cli/ingest.go
package cli

// Ingest command — chunk files (or stdin) into memories.

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/nexusriot/ai-houkai/internal/ingest"
	"github.com/nexusriot/ai-houkai/internal/memory"
	"github.com/spf13/cobra"
)

func newIngestCmd() *cobra.Command {
	var memType, source string
	var tags []string
	var importance float32
	var autoImportance, dryRun, yes bool
	var maxChars, minChars int

	cmd := &cobra.Command{
		Use:   "ingest [files...]",
		Short: "Split files into chunks and store each chunk as a memory",
		Long: `Split files into chunks and store each chunk as a memory.

Splits on blank lines, keeps markdown headings attached to their
paragraph, re-packs long paragraphs on sentence boundaries. Omit files
(or pass '-') to read stdin.`,
		RunE: func(cmd *cobra.Command, args []string) error {
			store := storeFromCtx(cmd.Context())
			cfg := cfgFromCtx(cmd.Context())

			// (label, text) per input.
			type input struct{ label, body string }
			var inputs []input
			if len(args) == 0 || (len(args) == 1 && args[0] == "-") {
				body, err := io.ReadAll(os.Stdin)
				if err != nil {
					return err
				}
				if strings.TrimSpace(string(body)) == "" {
					return fmt.Errorf("no input provided")
				}
				inputs = append(inputs, input{"stdin", string(body)})
			} else {
				for _, f := range args {
					data, err := os.ReadFile(f)
					if err != nil {
						return fmt.Errorf("%s not found", f)
					}
					inputs = append(inputs, input{filepath.Base(f), string(data)})
				}
			}

			explicit := cmd.Flags().Changed("importance")
			auto := autoImportance || (!explicit && cfg.DefaultImportance.Auto)

			type planned struct {
				label, chunk string
				imp          float32
			}
			var plan []planned
			for _, in := range inputs {
				for _, chunk := range ingest.ChunkText(in.body, maxChars, minChars) {
					var imp float32
					switch {
					case explicit:
						imp = importance
					case auto:
						imp = memory.ScoreImportance(chunk, memory.MemoryType(memType), tags)
					default:
						imp = cfg.DefaultImportance.Value
					}
					plan = append(plan, planned{in.label, chunk, imp})
				}
			}

			if len(plan) == 0 {
				fmt.Println("Nothing to ingest (all chunks below --min-chars?).")
				return nil
			}

			for i, p := range plan {
				firstLine := strings.SplitN(p.chunk, "\n", 2)[0]
				if len(firstLine) > 70 {
					firstLine = firstLine[:70]
				}
				fmt.Printf("  [%3d] %.2f  %s: %s\n", i+1, p.imp, p.label, firstLine)
			}
			fmt.Printf("\n%d chunk(s) from %d input(s).\n", len(plan), len(inputs))

			if dryRun {
				fmt.Println("Dry-run — nothing written.")
				return nil
			}

			if !yes && !Confirm(fmt.Sprintf("Store %d memories?", len(plan))) {
				fmt.Println("Aborted.")
				return nil
			}

			defer store.AsActor("import")()
			// One batched write — collapses N per-chunk encodes into
			// ceil(N/batch) (see MemoryStore.RememberMany). Ingesting raw
			// document chunks shouldn't trigger conflict management, so ignore.
			batch := make([]memory.RememberItem, len(plan))
			for i, p := range plan {
				src := source
				if src == "" {
					src = "ingest:" + p.label
				}
				batch[i] = memory.RememberItem{
					Text: p.chunk,
					RememberOpts: memory.RememberOpts{
						Type:       memory.MemoryType(memType),
						Tags:       tags,
						Importance: memory.Float32Ptr(p.imp),
						Source:     src,
					},
				}
			}
			if _, err := store.RememberMany(cmd.Context(), batch, 128, memory.PolicyIgnore); err != nil {
				return err
			}
			fmt.Printf("Stored %d memories.\n", len(plan))
			return nil
		},
	}
	cmd.Flags().StringVarP(&memType, "type", "t", "episodic", "episodic|semantic|procedural|feedback")
	cmd.Flags().StringSliceVarP(&tags, "tag", "g", nil, "Tag (repeatable)")
	// No -s shorthand: it would collide with the persistent --store flag.
	cmd.Flags().StringVar(&source, "source", "", "Source label; default ingest:<filename> (or ingest:stdin)")
	cmd.Flags().Float32VarP(&importance, "importance", "i", 0, "Importance 0.0–1.0")
	cmd.Flags().BoolVar(&autoImportance, "auto-importance", false, "Score each chunk heuristically")
	cmd.Flags().IntVar(&maxChars, "max-chars", 500, "Max chunk size")
	cmd.Flags().IntVar(&minChars, "min-chars", 30, "Drop chunks shorter than this")
	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "Show chunks without writing")
	cmd.Flags().BoolVarP(&yes, "yes", "y", false, "Skip confirmation")
	return cmd
}
